# Data Cleaning

### 1 Importing The Necessary Libraries

In [1]:
# importing the necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from datetime import datetime
pd.set_option("display.max_columns",None)
warnings.filterwarnings("ignore")

### 2 Loading The Dataset and Data Exploration

In [3]:
# Loading the dataset
df = pd.read_csv("C:\\Users\\yozil\\Desktop\\My projects\\4.0 Fraud_Detection\\data\\raw\\creditcard.csv")
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9       V10       V11       V12  \
0       0.462388  0.239599  0.098698  0.363787  0.090794 -0.551600 -0.617801   
1      -0.082361 -0.078803  0.085102 -0.255425 -0.166974  1.612727  1.065235   
2       1.800499  0.791461  0.247676 -1.514654  0.207643  0.624501  0.066084   
3       1.247203  0.237609  0.377436 -1.387024 -0.054952 -0.226487  0.178228   
4       0.095921  0.592941 -0.270533  0.817739  0.753074 -0.822843  0.538196   
...          ...       ...       ...       ...       ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  4.356170 -1.593105  2.711941   
284803  1.058415  0.024330  0.294869  0.584800 -0.975926 -0.150189  0.915802   
284804  3.031260 -0.296827  0.708417  0.432454 -0.484782  0.411614  0.063119   
284805  0.623708 -0.686180  0.679145  0.392087 -0.399126 -1.933849 -0.962886   
284806 -0.649617  1.577006 -0.414650  0.486180 -0.915427 -1.040458 -0.031513   

             V13       V14       V15       V16       V17       V18       V19  \
0      -0.991390 -0.311169  1.468177 -0.470401  0.207971  0.025791  0.403993   
1       0.489095 -0.143772  0.635558  0.463917 -0.114805 -0.183361 -0.145783   
2       0.717293 -0.165946  2.345865 -2.890083  1.109969 -0.121359 -2.261857   
3       0.507757 -0.287924 -0.631418 -1.059647 -0.684093  1.965775 -1.232622   
4       1.345852 -1.119670  0.175121 -0.451449 -0.237033 -0.038195  0.803487   
...          ...       ...       ...       ...       ...       ...       ...   
284802 -0.689256  4.626942 -0.924459  1.107641  1.991691  0.510632 -0.682920   
284803  1.214756 -0.675143  1.164931 -0.711757 -0.025693 -1.221179 -1.545556   
284804 -0.183699 -0.510602  1.329284  0.140716  0.313502  0.395652 -0.577252   
284805 -1.042082  0.449624  1.962563 -0.608577  0.509928  1.113981  2.897849   
284806 -0.188093 -0.084316  0.041333 -0.302620 -0.660377  0.167430 -0.256117   

             V20       V21       V22       V23       V24       V25       V26  \
0       0.251412 -0.018307  0.277838 -0.110474  0.066928  0.128539 -0.189115   
1      -0.069083 -0.225775 -0.638672  0.101288 -0.339846  0.167170  0.125895   
2       0.524980  0.247998  0.771679  0.909412 -0.689281 -0.327642 -0.139097   
3      -0.208038 -0.108300  0.005274 -0.190321 -1.175575  0.647376 -0.221929   
4       0.408542 -0.009431  0.798278 -0.137458  0.141267 -0.206010  0.502292   
...          ...       ...       ...       ...       ...       ...       ...   
284802  1.475829  0.213454  0.111864  1.014480 -0.509348  1.436807  0.250034   
284803  0.059616  0.214205  0.924384  0.012463 -1.016226 -0.606624 -0.395255   
284804  0.001396  0.232045  0.578229 -0.037501  0.640134  0.265745 -0.087371   
284805  0.127434  0.265245  0.800049 -0.163298  0.123205 -0.569159  0.546668   
284806  0.382948  0.261057  0.643078  0.376777  0.008797 -0.473649 -0.818267   

             V27       V28  Amount  Class  
0       0.133558 -0.021053  149.62      0  
1      -0.008983  0.014724    2.69      0  
2      -0.055353 -0.059752  378.66      0  
3       0.062723  0.061458  123.50      0  
4       0.219422  0.215153   69.99      0  


In [4]:
# displaying 3 sample records from our dataset
df.sample(3)

Time        V1        V2        V3        V4        V5        V6  \
78942    57783.0 -1.794857  1.124720  0.179232 -2.427335  0.196062 -0.969808   
220496  142197.0 -1.570301 -1.025119 -0.079181 -2.128794 -1.974789  0.010837   
255693  157351.0 -0.183503  1.309022 -0.619935 -0.584845  0.860050 -0.803773   

              V7        V8        V9       V10       V11       V12       V13  \
78942   0.579130  0.515615  0.199792 -1.432433  0.458249  0.815743 -0.504339   
220496  2.217667  0.267945  0.177605 -2.066574  0.617704  1.092652  0.136749   
255693  1.001235 -0.178526 -0.162635 -0.227197 -1.380778  0.630268  1.780221   

             V14       V15       V16       V17       V18       V19       V20  \
78942   1.016147  0.261781 -0.035981 -0.612508  0.405203  0.921359 -0.169165   
220496  0.351623 -1.129126  0.939407 -1.041480  0.252147 -0.819883  0.901600   
255693  0.130488  0.330590 -0.275940 -0.779962  0.208939  0.312572  0.048784   

             V21       V22       V23       V24       V25       V26       V27  \
78942  -0.297559 -1.002809 -0.336454 -0.554990  0.697254 -1.120106 -0.375169   
220496  0.311955 -0.259181  1.589709  0.004096 -0.947285 -0.084406 -0.183876   
255693  0.346834  1.350317 -0.232488 -0.632751 -0.651850 -0.220060  0.094302   

             V28  Amount  Class  
78942  -0.265168   24.70      0  
220496  0.184627  590.00      0  
255693  0.030090    0.75      0

In [7]:
# shape of the dataset
df.shape

(284807, 31)

In [8]:
# displaying general information of the dataframe including the memory usage
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [11]:
# column names
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [20]:
# number of categorical columns in the dataset
len(df.select_dtypes("object").columns)

0

As we can see from the result of the abobe cell we don't have any categorical variable in our dataset.

In [23]:
# number of numerical columns in the dataset
len(df.select_dtypes(["float64","int64"]).columns)

31

In [24]:
# numerical columns
df.select_dtypes(["float64","int64"]).columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [27]:
# statistical summary for numerical columns
df.describe().round(2)

Time         V1         V2         V3         V4         V5  \
count  284807.00  284807.00  284807.00  284807.00  284807.00  284807.00   
mean    94813.86       0.00       0.00      -0.00       0.00       0.00   
std     47488.15       1.96       1.65       1.52       1.42       1.38   
min         0.00     -56.41     -72.72     -48.33      -5.68    -113.74   
25%     54201.50      -0.92      -0.60      -0.89      -0.85      -0.69   
50%     84692.00       0.02       0.07       0.18      -0.02      -0.05   
75%    139320.50       1.32       0.80       1.03       0.74       0.61   
max    172792.00       2.45      22.06       9.38      16.88      34.80   

              V6         V7         V8         V9        V10        V11  \
count  284807.00  284807.00  284807.00  284807.00  284807.00  284807.00   
mean        0.00      -0.00       0.00      -0.00       0.00       0.00   
std         1.33       1.24       1.19       1.10       1.09       1.02   
min       -26.16     -43.56     -73.22     -13.43     -24.59      -4.80   
25%        -0.77      -0.55      -0.21      -0.64      -0.54      -0.76   
50%        -0.27       0.04       0.02      -0.05      -0.09      -0.03   
75%         0.40       0.57       0.33       0.60       0.45       0.74   
max        73.30     120.59      20.01      15.59      23.75      12.02   

             V12        V13        V14        V15        V16        V17  \
count  284807.00  284807.00  284807.00  284807.00  284807.00  284807.00   
mean       -0.00       0.00       0.00       0.00       0.00      -0.00   
std         1.00       1.00       0.96       0.92       0.88       0.85   
min       -18.68      -5.79     -19.21      -4.50     -14.13     -25.16   
25%        -0.41      -0.65      -0.43      -0.58      -0.47      -0.48   
50%         0.14      -0.01       0.05       0.05       0.07      -0.07   
75%         0.62       0.66       0.49       0.65       0.52       0.40   
max         7.85       7.13      10.53       8.88      17.32       9.25   

             V18        V19        V20        V21        V22        V23  \
count  284807.00  284807.00  284807.00  284807.00  284807.00  284807.00   
mean        0.00       0.00       0.00       0.00      -0.00       0.00   
std         0.84       0.81       0.77       0.73       0.73       0.62   
min        -9.50      -7.21     -54.50     -34.83     -10.93     -44.81   
25%        -0.50      -0.46      -0.21      -0.23      -0.54      -0.16   
50%        -0.00       0.00      -0.06      -0.03       0.01      -0.01   
75%         0.50       0.46       0.13       0.19       0.53       0.15   
max         5.04       5.59      39.42      27.20      10.50      22.53   

             V24        V25        V26        V27        V28     Amount  \
count  284807.00  284807.00  284807.00  284807.00  284807.00  284807.00   
mean        0.00       0.00       0.00      -0.00      -0.00      88.35   
std         0.61       0.52       0.48       0.40       0.33     250.12   
min        -2.84     -10.30      -2.60     -22.57     -15.43       0.00   
25%        -0.35      -0.32      -0.33      -0.07      -0.05       5.60   
50%         0.04       0.02      -0.05       0.00       0.01      22.00   
75%         0.44       0.35       0.24       0.09       0.08      77.16   
max         4.58       7.52       3.52      31.61      33.85   25691.16   

           Class  
count  284807.00  
mean        0.00  
std         0.04  
min         0.00  
25%         0.00  
50%         0.00  
75%         0.00  
max         1.00

## 3 Data Cleaning

##### 3.1 Standardizing Column Names

In [28]:
# defining a function to standardize column names
def column_maker(column_name):
    return column_name.str.title()

In [31]:
# applying column_maker function
df.columns = column_maker(df.columns)

In [33]:
# checking our column names
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

##### 3.2 Removing Duplicated Records (if any)

In [36]:
# first let's check the existance of duplicated records
df.duplicated().any()

True

the above result tells us the existance of duplicated records in our dataframe

In [37]:
# let's see how many records are duplicated
df.duplicated().sum()

1081

we have 1081 duplicated results in our data frame

In [38]:
# now let's remove the duplicated records from our dataframe
df.drop_duplicates(inplace=True)

In [39]:
# now let's check the shape of our new dataframe
df.shape

(283726, 31)

In [41]:
# since we have removed records from our dataframe, we have to fix the index
df.reset_index()

index      Time         V1         V2        V3        V4        V5  \
0            0       0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            1       0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            2       1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            3       1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            4       2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...        ...       ...        ...        ...       ...       ...       ...   
283721  284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
283722  284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
283723  284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
283724  284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
283725  284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9       V10       V11       V12  \
0       0.462388  0.239599  0.098698  0.363787  0.090794 -0.551600 -0.617801   
1      -0.082361 -0.078803  0.085102 -0.255425 -0.166974  1.612727  1.065235   
2       1.800499  0.791461  0.247676 -1.514654  0.207643  0.624501  0.066084   
3       1.247203  0.237609  0.377436 -1.387024 -0.054952 -0.226487  0.178228   
4       0.095921  0.592941 -0.270533  0.817739  0.753074 -0.822843  0.538196   
...          ...       ...       ...       ...       ...       ...       ...   
283721 -2.606837 -4.918215  7.305334  1.914428  4.356170 -1.593105  2.711941   
283722  1.058415  0.024330  0.294869  0.584800 -0.975926 -0.150189  0.915802   
283723  3.031260 -0.296827  0.708417  0.432454 -0.484782  0.411614  0.063119   
283724  0.623708 -0.686180  0.679145  0.392087 -0.399126 -1.933849 -0.962886   
283725 -0.649617  1.577006 -0.414650  0.486180 -0.915427 -1.040458 -0.031513   

             V13       V14       V15       V16       V17       V18       V19  \
0      -0.991390 -0.311169  1.468177 -0.470401  0.207971  0.025791  0.403993   
1       0.489095 -0.143772  0.635558  0.463917 -0.114805 -0.183361 -0.145783   
2       0.717293 -0.165946  2.345865 -2.890083  1.109969 -0.121359 -2.261857   
3       0.507757 -0.287924 -0.631418 -1.059647 -0.684093  1.965775 -1.232622   
4       1.345852 -1.119670  0.175121 -0.451449 -0.237033 -0.038195  0.803487   
...          ...       ...       ...       ...       ...       ...       ...   
283721 -0.689256  4.626942 -0.924459  1.107641  1.991691  0.510632 -0.682920   
283722  1.214756 -0.675143  1.164931 -0.711757 -0.025693 -1.221179 -1.545556   
283723 -0.183699 -0.510602  1.329284  0.140716  0.313502  0.395652 -0.577252   
283724 -1.042082  0.449624  1.962563 -0.608577  0.509928  1.113981  2.897849   
283725 -0.188093 -0.084316  0.041333 -0.302620 -0.660377  0.167430 -0.256117   

             V20       V21       V22       V23       V24       V25       V26  \
0       0.251412 -0.018307  0.277838 -0.110474  0.066928  0.128539 -0.189115   
1      -0.069083 -0.225775 -0.638672  0.101288 -0.339846  0.167170  0.125895   
2       0.524980  0.247998  0.771679  0.909412 -0.689281 -0.327642 -0.139097   
3      -0.208038 -0.108300  0.005274 -0.190321 -1.175575  0.647376 -0.221929   
4       0.408542 -0.009431  0.798278 -0.137458  0.141267 -0.206010  0.502292   
...          ...       ...       ...       ...       ...       ...       ...   
283721  1.475829  0.213454  0.111864  1.014480 -0.509348  1.436807  0.250034   
283722  0.059616  0.214205  0.924384  0.012463 -1.016226 -0.606624 -0.395255   
283723  0.001396  0.232045  0.578229 -0.037501  0.640134  0.265745 -0.087371   
283724  0.127434  0.265245  0.800049 -0.163298  0.123205 -0.569159  0.546668   
283725  0.382948  0.261057  0.643078  0.376777  0.008797 -0.473649 -0.818267   

             V27       V28  Amount  Class  
0       0.133558 -0.021053  149.62      0  
1      -0.008983  0.014724    2.69      0  
2      -0.055353 -0.059752  378.6

In [42]:
# now let's remove the index column from our dataframe
df.drop("index", axis = 1, inplace = True)

KeyError: "['index'] not found in axis"

In [ ]:
# now let'